### ISSUE TITLE
[BUG] umbrella issue: sporadic failures of monotonicity in `test_predict_quantiles`

### ISSUE THREAD

**Issue by fkiraly** [sktime/sktime#4431]
Multiple forecasters produce, on occasion, quantile forecasts that are not monotonic:

* `VAR` https://github.com/sktime/sktime/issues/4420
* `Prophet`

This is sporadic and happens in a small percentage of cases.

A commonality may be doing Bayesian sampling internally, e.g., sampling is done independently for the different `alpha` or `coverage` values, so the approximated posteriors may end up crossing (while the real posteriors don't)

**Comment by fkiraly** [comment]
FYI @yarnabrina

**Comment by yarnabrina** [comment]
> should we allow this to happen

A very strict way will be to not allow at all, and let it fail. A slightly less strict (and debatable) way can be to run once again and then fail or return based on monotonicity.

> should we force-sort returned prediction intervals

If I am to express my opinion, I do not want this. Even if we allow reverse quantiles, one possibly still justify the reasons causing it and the associated likelihoods, as you did above. I'll prefer un-sorted quantiles with a warning (generated if quantiles != sorted(quantiles) etc.), over force-sorted quantiles. Or, this can be left to users as well, e.g. `warn_on_monotonicity=True`.

> A commonality may be doing Bayesian/MC sampling internally, e.g., sampling is done independently for the different alpha or coverage values, so the approximated posteriors may end up crossing (while the real posteriors don't)

I know some of the probabilistic forecast packages like `gluonts` predict this way, but is it done for all the models? Or is this just an example?

May be we should attempt (somehow) to generate at least one occurrence of unsorted quantile predictions for underlying original estimators. That would at least ensure that nothing in `sktime` is causing it.

**Comment by fkiraly** [comment]
> If I am to express my opinion, I do not want this.

I would agree with you there. "if you build a framework, don't police the instance" would be the design principle, i.e., giving implementers freedom to move within the framework.
If they want crossing quantiles, so be it (perhaps there is a reason why one would like this).

> May be we should attempt (somehow) to generate at least one occurrence of unsorted quantile predictions for underlying original estimators. 

Agreed. We should check whether it's coming from upstream.

> A likely stupid question: is it correct to use predict_interval for predict_quantiles? The former says this interval will contain the actual value {coverage}% cases, and the latter says {alpha} is the chance for actual value to be less than this value. Do these necessarily correspond to each other, across any types of skewness?

In-principle, estimators can produce intervals and quantiles that are inconsistent (it is not tested or part of the contract).
So, *contract-wise*, these do not need to correspond currently.

However, as far as I know, in all estimators in `sktime`, the intervals are symmetric quantile intervals, so in the *implemented cases*, they always 1:1 correspond (via the formula that c coverage interval is the interval that starts at the 1/2-1/2 c quantile and ends at the 1/2 + 1/2 c quantile).

**Comment by fkiraly** [comment]
> Or, this can be left to users as well, e.g. `warn_on_monotonicity=True`.

This could be a configuration field for forecasters (for `set_config`), with a default = True.

Expanding on the "left to users" - perhaps the sorting itself could *also be left to users*?

There could be a simple wraller, e.g., `SortQuantiles` or `FixQuantileCrossing` (with different methods/strategies), in which you stick any forecaster, and it ensures that the quantiles are sorted. This could be done in different ways, naive sorting, or Monte Carlo based, or sophisticated ways such as density estimation.

**Comment by yarnabrina** [comment]
Yes, giving users the option to sort makes sense 👍. Quite a few `pandas`
functions have options for `raise`, `ignore`, `coerce`, etc. on error.
Similarly `sktime` can choose to let it fail, do nothing or sort.

**Comment by fkiraly** [comment]
hm, what would be your design approach, @yarnabrina?
A config flag with `raise`, `ignore`, `coerce`? Default `ignore`?

**Comment by yarnabrina** [comment]
Exactly, except may be I personally would have preferred `raise` as default. I'll give my (possibly biased) reasoning below.

If most of the times the results are sorted, similar to the common expectation, I as a user is less likely to check for monotonicity in scheduled workflows. I'd probably set up some regular workflow to take the results and pass them to store in databases or for plotting in some dashboards. There will be some try-except to handle in case something goes wrong during train/predict, but not for quantile ordering. As a result, if order is actually reverse for a particular case, and `ignore` is applied, the plots are going to be messed up, and no one would know that until it's too late. But if it's `raise`, then ideally it'd notify someone that something's wrong and needs to be validated manually and they can then decide what to do.

I understand that `coerce` for auto-sorting sounds like a good option to handle the scenario I just said above, but that seems wrong personally and I'd prefer the framework to do it only if user asked for it explicitly.

**Comment by fkiraly** [comment]
I see, I agree with your reasoning on both points after reading, @yarnabrina.

There is one drawback to `raise` though: the estimator breaks on occasion for sure, even in a case where the regular workflow would not require non-crossing quantiles. Let's say the probability for crossing uantiles is 1% in the production state. Then, you can imagine a case where the user doesn't have a catch for the failure case, and has run 50 test runs where everything worked, so has some degree of trust in the workflow. But then it breaks in case 51.

For me, the discussion highlights two things:

* `warn` should be an option, too - perhaps even the default, as it does have neither of the problems while satisfying your requirement of notifying the user
* should there be a tag that tells the user "this can in-principle happen"? E.g., this is an estimator which may produce crossing quantiles

**Comment by yarnabrina** [comment]
> should there be a tag that tells the user "this can in-principle happen"? E.g., this is an estimator which may produce crossing quantiles

Sounds nice to me, but how do you propose to identify such estimators? Run all estimators a lot of times and hope the one passes will be good forever?

> it does have neither of the problems while satisfying your requirement of notifying the user

Agreed, but just making an observation here, a lot of estimators throw a lot of convergence warnings frequently, so it's not vey uncommon to disable with `warnings.filterwarnings("ignore")`. But an argument can be made that user explicitly disabled that on their own choice.

**Comment by fkiraly** [comment]
> Sounds nice to me, but how do you propose to identify such estimators? Run all estimators a lot of times and hope the one passes will be good forever?

That's a fair point. Sounds tricky, and casts some doubt on the usefulness of the tag.

How about this: default is "it never happens", until it happens, and then we set the tag to `True`.
For a start, we can add the estimators at the top of this issue.
Going forward, we ensure the test fails if the tag says "it never happens" but it happens, so we would successively identify all.

> But an argument can be made that user explicitly disabled that on their own choice.

Yes.

> but just making an observation here, a lot of estimators throw a lot of convergence warnings frequently

Hm, is that also a case for a config field? Disabling this per estimator?
